### PR TITLE
spatial audio: isolate PannerNode rotation

### PR DESCRIPTION
**Pull request by kanyesthaker**
## Description of Changes
Fixes a bug where a user's audio would suddenly become quieter or louder after rapidly muting/unmuting due to undesired PannerNode rotations. Changing state would sometimes cause multiple consescutive rotations, causing the node to point the wrong way. This PR ensures it only happens once.

## Testing Checklists
If your change adds critical functionality that should be tested on every PR, edit ``pull_requeste_template.md`` and add it to the checklist.
### Basic Author Checklist 
The PR author should verify that the following are fully functional.
- [x] With 1 user:
  - [x] The world loads, the user can move/look around, and physics behave properly
  - [x] User can pause their client, mute/unmute themselves, and change their name
- [x] With 2 users:
  - [x] A second user can join the same room, and both users can see each others model/video and hear each others audio
  - [x] State changes (name changes, A/V toggles) propagate between peers as expected

### PR Specific Author Checklist
What did you do to test your change in particular? For example, if your changes touches A/V functionality, how did you stress-test this part of the product?
Before:
[Screencast from 12-12-2023 11:40:22 PM.webm](https://github.com/IrisHub/SpaceBar/assets/31911175/45426a26-b0ad-4bf0-a8f0-be1e29ab4923)

After:
https://github.com/IrisHub/SpaceBar/assets/31911175/04cec2bd-a799-485f-b299-6a3cd0fd45a6

### Advanced Author Checklist
These are more advanced checks that might be helpful for some PRs. Completing these is at the discretion of the author and reviewer.
- [ ] With 2 users, memory stays within a reasonable interval over a 5 minute period
- [ ] With 2 users, the server remains healthy after an extended amount of time in the world
- [ ] With 3+ users, the server and client behave in a manner consistent with 2 peers

### Basic Reviewer Checklist
The reviewer of this PR should load SpaceBar off this branch and ensure the following work as expected:
- [x] 2 user can join the world and see each other's model/video, and hear each other's audio
- [x] Movement and collisions behave as expected
- [x] Changing some state (name, microphone, video) for one user propagates correctly to the other user

### CODE REVIEW

File: src/scene/players/PeerPlayerMediaProvider.tsx [hunk 1]
```diff
@@ -94,27 +94,31 @@ const PeerPlayerMediaProvider = ({
...
-        const helper = new PositionalAudioHelper(audioRef.current, 3, 10, 5);
-        audioRef.current.add(helper);
-      }
       log.debug(`${Date.now()}: Set audio from peer stream`);
     }
   }, [peerStream, playerMediaMetadataMap.get(peerId)?.isAudioEnabled]);
 
+  // Add properties to the audioRef to make audio work spatially
+  // In a separate useEffect with no deps to make sure it only happens once
+  useEffect(() => {
```
Comment (samjgorman) [inline src/scene/players/PeerPlayerMediaProvider.tsx:103]: For readability I'd move this side effect before the later side effect that updates audio on change in `peerStream`. Recommending as this code is responsible for the initialization of properties for the `audioRef`

**Review by samjgorman — APPROVED** [review]

File: src/scene/players/PeerPlayerMediaProvider.tsx [hunk 2]
```diff
@@ -94,27 +94,31 @@ const PeerPlayerMediaProvider = ({
...
+    // Multiplying quaternions is equivalent to composing them, meaning this multiplication is like
+    // Rotating the PannerNode 180 degrees in its own local coordinate system
+    const inverseYQuat = new Quaternion(0, 1, 0, 0);
+    audioRef.current.setRotationFromQuaternion(
+      audioRef.current.quaternion.multiply(inverseYQuat),
+    );
+    audioRef.current.setDirectionalCone(...AudioConstants.DirectionalCone); // Directional cone outside of which audio is muted
+
+    if (process.env.NEXT_PUBLIC_SANDBOX_FLAG === 'true') {
+      const helper = new PositionalAudioHelper(audioRef.current, 3, 10, 5);
```
[the remark this answers is not in the document]
Comment (kanyesthaker) [reply src/scene/players/PeerPlayerMediaProvider.tsx:117]: they're only used for debugging, not functional

File: src/scene/players/PeerPlayerMediaProvider.tsx [hunk 3]
```diff
@@ -94,27 +94,31 @@ const PeerPlayerMediaProvider = ({
...
-
-      if (process.env.NEXT_PUBLIC_SANDBOX_FLAG === 'true') {
-        const helper = new PositionalAudioHelper(audioRef.current, 3, 10, 5);
-        audioRef.current.add(helper);
-      }
       log.debug(`${Date.now()}: Set audio from peer stream`);
     }
   }, [peerStream, playerMediaMetadataMap.get(peerId)?.isAudioEnabled]);
 
+  // Add properties to the audioRef to make audio work spatially
```
[the remark this answers is not in the document]
Comment (kanyesthaker) [reply src/scene/players/PeerPlayerMediaProvider.tsx:101]: good idea!

File: src/scene/players/PeerPlayerMediaProvider.tsx (hunk 1 above)
Comment (kanyesthaker) [reply src/scene/players/PeerPlayerMediaProvider.tsx:103]: sure!

File: src/scene/players/PeerPlayerMediaProvider.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (samjgorman) [reply src/scene/players/PeerPlayerMediaProvider.tsx]: I'd still want to see constants -- I have no idea what these numbers represent so would have a hard time effectively using this while debugging

**Pull request merged**